### DISCUSSION TITLE
Internal sensors missing from ESP32-C3 port of the ESP library?

### DISCUSSION THREAD

**Discussion by Ajb2k3** [micropython/micropython#10443] [ESP32] [answered]
In the Port of the DOWDQ6 and Pico D4 variants of the ESP32 the esp library contains functions for the internal hall and temperature sensor however, on the ESP32-C3 Port these functions are missing from the esp lib. Were they moved or just not added?

**Comment by fivdi** [comment, marked as answer]
The ESP32-C3 has a temperature sensor but it doesn't have a hall sensor. At least there is no mention of a hall sensor in ESP32-C3 datasheet.

If I understand the source code correctly, functions `esp32.hall_sensor()` for reading the internal hall sensor and function `esp32.raw_temperature()` for reading the internal temperature are only available for the ESP32 but not for the ESP32-C3, ESP32-S2 or ESP32-S3. The code is [here](https://github.com/micropython/micropython/blob/v1.19.1/ports/esp32/modesp32.c#L144-L170) and is only added if `CONFIG_IDF_TARGET_ESP32` is defined. The code is not added when `CONFIG_IDF_TARGET_ESP32C3` is defined (which is the corresponding define for the ESP32-C3.)

**Reply by Ajb2k3** [reply]
Your right, it is just the temperature sensor (I was reading another data sheet) 
But if the data sheet says there is an internal temperature sensor, why is it not available in the port?
Sorry but I'm struggling to understand none python code due to learning difficulties.

**Reply by fivdi** [reply]
>  But if the data sheet says there is an internal temperature sensor, why is it not available in the port?

I guess nobody had the time to add and test the required code. (I'm not sure if the ESP32 and ESP32-C3 use the same code...)

**Reply by scruss** [reply]
Also, the Hall sensor (on regular ESP32s only, not C or S variants) is no longer supported on IDF ≥ 5

**Comment by MATTYGILO** [comment]
I have CONFIG_IDF_TARGET_ESP32S3=y defined however it isn't present in my code
<img width="544" alt="Screenshot 2023-11-10 at 21 28 34" src="https://github.com/micropython/micropython/assets/34808802/f70c63ed-f5ee-45b7-9672-440731050bf2">

**Reply by fivdi** [reply]
As mentioned above, `raw_temperature` is only available when `CONFIG_IDF_TARGET_ESP32` is defined. You are defining `CONFIG_IDF_TARGET_ESP32S3` which is not the same thing. I guess you are using an ESP32S3 and not an ESP32. See [here](https://github.com/micropython/micropython/blob/v1.21.0/ports/esp32/modesp32.c#L151-L171).

**Reply by MATTYGILO** [reply]
@fivdi Yes I am using a esp32s3, however esp32s3 also has a raw temp sensor, so it should be available for it too

**Reply by sophiedegran** [reply]
For me most of ESP32S3 have no more internal sensor temperature (it's a long history..) Espressif maintain the library because for some production they active it.. Try in C you will see if it's active or not ...

**Comment by rini711** [comment]
I want to add raw_temperature for ESP32C3 , How can we add it .I tried to use ESP32 code but getting below error
![image](https://github.com/micropython/micropython/assets/144786971/dcdfcc50-e7cc-4cdd-bb80-6ad3a5749111)

**Reply by sophiedegran** [reply]
sens_reg.h is in esp32s3 dir in idf 5.0.2
(components\soc)
but as I tell you on my previous post test before on pure C with arduino IDE ...

**Reply by ricksorensen** [reply]
I modified esp32mod.c and got a initial read of the internal temperature using the v5 API ...  
[https://docs.espressif.com/projects/esp-idf/en/latest/esp32c3/api-reference/peripherals/temp_sensor.html#:~:text=The%20ESP32%2DC3%20has%20a,compensate%20for%20the%20temperature%20measurement.](https://docs.espressif.com/projects/esp-idf/en/latest/esp32c3/api-reference/peripherals/temp_sensor.html#:~:text=The%20ESP32%2DC3%20has%20a,compensate%20for%20the%20temperature%20measurement.)
After the conditional compile option  for the esp32 (circa line 172) I added a brute force copy from the ESP-IDF example:
```
#if CONFIG_IDF_TARGET_ESP32C3
#include "driver/temperature_sensor.h"
STATIC mp_obj_t esp32_raw_temperature(void) {
  temperature_sensor_config_t temp_sensor_config = TEMPERATURE_SENSOR_CONFIG_DEFAULT(10,50);
  temperature_sensor_handle_t temp_sensor=NULL;
  temperature_sensor_install(&temp_sensor_config, &temp_sensor);
  temperature_sensor_enable(temp_sensor);
  float tvalue;
  temperature_sensor_get_celsius(temp_sensor, &tvalue);
  temperature_sensor_disable(temp_sensor);
  temperature_sensor_uninstall(temp_sensor);
  return mp_obj_new_int((int)(tvalue+0.5));
}
STATIC MP_DEFINE_CONST_FUN_OBJ_0(esp32_raw_temperature_obj, esp32_raw_temperature);

#endif
```
Then one needs to add this definition for the esp32 module actions after the ESP32 definition (after oriiginal line 205):
```
    #if CONFIG_IDF_TARGET_ESP32C3
    { MP_ROM_QSTR(MP_QSTR_raw_temperature), MP_ROM_PTR(&esp32_raw_temperature_obj) },
    #endif
```
This seemed to work for my ESP32C3 (SEEED XIAO) - but I have not tested this thoroughly yet.  I also wonder if this fix will work with any ESP32 board- the API seems general for all.

Caveats: 

- Hacky read of float then convert to int ...  
- I wonder if there is a more direct sensor read option?  (Remember- the 32C is a RISC-V processor not eXtensa as the other variants) 
- `installation` could be supported directly, instead of every call install/enable/read/disable/uninstall.
- no error handling in this demonstration
-

**Reply by sophiedegran** [reply]
Many thanks @ricksorensen I will look at that but don't forget that Espressif enable/ disable these internal sensors at the line production... What about to put a simple dht11 outside that cost nothing !..

**Reply by robert-hh** [reply]
The internal temperature sensor is not really useful to tell the ESP32-outside temperature. Even with external sensors you have to  take into account, that reading the value too frequent may heat up the sensor.

**Comment by sophiedegran** [comment]
You right robert but as you're explain in july 2019 (old forum) the max temperature is 125*C so 50*C-60*C it's Ok! for Esp32 chipset.. So this value is to take with precaution... That's why for me it's more useable to have a sensor in the case and not to monitor the chipset ...

**Comment by rini711** [comment]
@ricksorensen and @Everyone , I have added the code in modesp32.c file and build the firmware, Now I am able to read temperature only first time after hard resetting device after that continuously 0 value is coming.
Below is the screenshot for reference.

Also, the temperature value which is reading first time after hard resetting device is more than the actual value.

Is there any way to fix both the issues??

<img width="946" alt="image" src="https://github.com/micropython/micropython/assets/144786971/c308db58-e3fe-4619-93d0-05933e4f7019">


CODE SCREENSHOT

![image](https://github.com/micropython/micropython/assets/144786971/2f20fc7b-21d3-490d-8567-242f56752961)

![image](https://github.com/micropython/micropython/assets/144786971/f5592066-1c1b-4b59-879d-ca946ade9f2c)

**Reply by robert-hh** [reply]
You could try to install& enable the sensor only once at the first call and keep it installed and enabled after that.
About the value: Which value did you expect? The sensor tells the internal temperature of the ESP32 chip, which is always higher than the temperature of the environment. For the internal temperature of the ESP32 39°C seems reasonable.

**Comment by rini711** [comment]
I want to read internal temperature of the ESP32 chip and it is coming more than the Actual Value. 

Please find below screenshot for micropython :
<img width="340" alt="image" src="https://github.com/micropython/micropython/assets/144786971/b5ca434a-50bb-4ba1-8ee7-c0b8cd0e6dc3">



When I tried same thing using Embedded C below code, Right value of temperature  I am getting.
<img width="660" alt="Screenshot 2023-12-12 at 2 15 32 PM" src="https://github.com/micropython/micropython/assets/144786971/57190479-45bd-4ff8-8fb1-7b862dd055c5">

#include "freertos/FreeRTOS.h"
#include "freertos/task.h"
#include "esp_log.h"
#include "driver/temperature_sensor.h"

static const char *TAG = "Device";

void app_main(void)
{
    ESP_LOGI(TAG, "Install temperature sensor, expected temp ranger range: 10~50 ℃");
    temperature_sensor_handle_t temp_sensor = NULL;
    temperature_sensor_config_t temp_sensor_config = TEMPERATURE_SENSOR_CONFIG_DEFAULT(10, 50);
    ESP_ERROR_CHECK(temperature_sensor_install(&temp_sensor_config, &temp_sensor));

    ESP_LOGI(TAG, "Enable temperature sensor");
    ESP_ERROR_CHECK(temperature_sensor_enable(temp_sensor));

    ESP_LOGI(TAG, "Read temperature");
    //int cnt = 20;
    float tsens_value;
    while (1) {
        ESP_ERROR_CHECK(temperature_sensor_get_celsius(temp_sensor, &tsens_value));
        ESP_LOGI(TAG, "Temperature value %.02f ℃", tsens_value);
        vTaskDelay(pdMS_TO_TICKS(1000));
    }
}

How to get correct temperature value after modifying modesp32.c file??

**Reply by robert-hh** [reply]
How do you know the right temperature in the respective configuration of the ESP32? You have two code packages for the ESP32 which run the device in different mode. So both values are probably correct.

**Reply by rini711** [reply]
Yes. The embedded c code response varies when we blow hot air and also I could relate it to the temperature I measured using thermocouple. On the other hand, the response from micropython is constant; unable to relate it with my thermocouple measurement.

**Reply by robert-hh** [reply]
Does the results of the C code get beyond 40°C when you heat up the device? And does the value returned my MicroPython decrease when you cool down the ESP32, like with cooling spray? Since in both cases you code just calls the esp-idf API with the same sequence, the difference in results must be related to setting up or some internal limits. The only difference I see from your code snippets above is, that teh C code waits between temperature readings and does not re-configure the sensor, while the Python code does not wait, but re-configures.

**Reply by ricksorensen** [reply]
Here's a not well controlled, not very precise trial   
SEEED XIAO ESP32C3 with raw_temperature() code added to `modesp32.c` .. as shown above.- install, enable, read, disable, uninstall on each read.
```>>> import time
>>> import math
>>> while True:
...     print(" st: ",esp32.raw_temperature())
...     y=1.0
...     for x in range(100):
...         y=y*(math.sin(x/21.0)*math.sin(x/21.0)+math.cos(x/21.0)*math.cos(x/21.0))
...     print(" aft: ",esp32.raw_temperature(), " ",y)
...     time.sleep(1)
... 
 st:  35
 aft:  38   0.9999989
 st:  38
 aft:  37   0.9999989
 st:  38
 aft:  38   0.9999989
 st:  38
 aft:  38   0.9999989
 st:  38
 aft:  37   0.9999989
 st:  38
 aft:  38   0.9999989
 st:  38
 aft:  38   0.9999989
 st:  38
 aft:  38   0.9999989
 st:  38
 aft:  37   0.9999989
 st:  38
 aft:  38   0.9999989
 st:  38
 aft:  38   0.9999989
 st:  38
 aft:  37   0.9999989
 st:  38
 aft:  38   0.9999989
 st:  38
 aft:  38   0.9999989
 st:  38
 aft:  37   0.9999989
 st:  38
 aft:  38   0.9999989
 st:  38
 aft:  38   0.9999989
 st:  38
 aft:  38   0.9999989
 st:  38
 aft:  38   0.9999989
 st:  38
 aft:  38   0.9999989
 st:  38
 aft:  38   0.9999989
 st:  38
 aft:  38   0.9999989
 st:  39
 aft:  38   0.9999989
 st:  39                                # started using hair dryer (old, not all that hot air coming out)
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  40   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  40   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  40   0.9999989
 st:  39
 aft:  40   0.9999989
 st:  39
 aft:  40   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  40   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  40   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  39   0.9999989
 st:  39
 aft:  40   0.9999989
Traceback (most recent call last):
  File "<stdin>", line 7, in <module>
KeyboardInterrupt: 
>>> 

```
Temp did not vary much- similar results with no delay at end of work loop.

**Reply by robert-hh** [reply]
So I grabbed en ESP32C3, my Hot Air solderign station and a can of cooling spray and tested. The results were erratic. Only the first value after a reset seemed fine. Then I changed the code into:
```
#if CONFIG_IDF_TARGET_ESP32C3
#include "driver/temperature_sensor.h"

STATIC mp_obj_t esp32_raw_temperature(void) {
    static temperature_sensor_handle_t temp_sensor = NULL;
    if (temp_sensor == NULL) {
        temperature_sensor_config_t temp_sensor_config = TEMPERATURE_SENSOR_CONFIG_DEFAULT(20, 100);
        ESP_ERROR_CHECK(temperature_sensor_install(&temp_sensor_config, &temp_sensor));
        ESP_ERROR_CHECK(temperature_sensor_enable(temp_sensor));
    }
    float tvalue;
    ESP_ERROR_CHECK(temperature_sensor_get_celsius(temp_sensor, &tvalue));
    return mp_obj_new_int((int)(tvalue+0.5));
}
STATIC MP_DEFINE_CONST_FUN_OBJ_0(esp32_raw_temperature_obj, esp32_raw_temperature);

#endif
```
That works. I can heat up the device and cool it down:
```
29  # Value after reset & programming. WiFi is off
29
29
29
29
31   # start with the hot air (set to 100°C)
33
35
38
40
41
43
44
45
46
47
48
48
48
48
47
47
47
46
48
50
53
55
56
58
60
61
62
63
63 # remove the hot air
62 
60
60
59
58
57
56
56   # Apply cooling spray
21
13
15
16
17
18
7
-9
-12
-11
-8
-6
-4
-2
-1
0
1
2
3
4
5
6
6
7
8
10
12
14
15
16
16
16
17
18 # Apply hot air again
18
18
19
20
20
20
20
22
23
24
24
24
25
29
40
44
43
42
41
41
40
```

**Reply by ricksorensen** [reply]
Thanks -  so it looks like the disable,uninstall sequence affects the performance.  I originally did the uninstall because the espressif indicated that saved power ... maybe a separate call function?

**Reply by robert-hh** [reply]
Yes. It could be a separate call.   I also tried to just run install once and keep enable/disable. But that did not work.

Current consumption before call to esp32.raw_temperature(): 13 - 14mA
Current consumption after call to esp32.raw_temperature(): 15mA
After starting WIFi: avg. 25mA.

Side note: It's winter here. I can cool down the MCU with my finger from 33°C to 30°C.

**Reply by robert-hh** [reply]
Tried again including enable/disable. Now it works, and the current consumption stays low. Changed code:
```
#if CONFIG_IDF_TARGET_ESP32C3
#include "driver/temperature_sensor.h"

STATIC mp_obj_t esp32_raw_temperature(void) {
    static temperature_sensor_handle_t temp_sensor = NULL;
    if (temp_sensor == NULL) {
        temperature_sensor_config_t temp_sensor_config = TEMPERATURE_SENSOR_CONFIG_DEFAULT(20, 100);
        ESP_ERROR_CHECK(temperature_sensor_install(&temp_sensor_config, &temp_sensor));
    }
    ESP_ERROR_CHECK(temperature_sensor_enable(temp_sensor));
    float tvalue;
    ESP_ERROR_CHECK(temperature_sensor_get_celsius(temp_sensor, &tvalue));
    ESP_ERROR_CHECK(temperature_sensor_disable(temp_sensor));
    return mp_obj_new_int((int)(tvalue+0.5));
}
STATIC MP_DEFINE_CONST_FUN_OBJ_0(esp32_raw_temperature_obj, esp32_raw_temperature);

#endif
```

**Reply by robert-hh** [reply]
I tried to return a float instead of the int, but that's not worth the effort. Testing it, the digit behind the dot was always 7. The test result above had a constant .2 for the values. The only addition to the function could be adding an optional argument(s) for the range.
```
27.7
27.7
28.7
28.7
25.7
23.7
22.7
21.7
21.7
21.7
21.7
21.7
21.7
21.7
22.7
22.7
23.7
23.7
23.7
23.7
24.7
24.7
24.7
25.7
25.7
25.7
```

**Reply by robert-hh** [reply]
Adding optional arguments to set the range's min and max, the code would look like:
```
STATIC mp_obj_t esp32_raw_temperature(size_t n_args, const mp_obj_t *args) {
    static temperature_sensor_handle_t temp_sensor = NULL;
    if (temp_sensor == NULL) {
        temperature_sensor_config_t temp_sensor_config = TEMPERATURE_SENSOR_CONFIG_DEFAULT(20, 100);
        if (n_args > 0) {
            temp_sensor_config.range_min = mp_obj_get_int(args[0]);
        }
        if (n_args > 1) {
            temp_sensor_config.range_max = mp_obj_get_int(args[1]);
        }
        ESP_ERROR_CHECK(temperature_sensor_install(&temp_sensor_config, &temp_sensor));
    }
    ESP_ERROR_CHECK(temperature_sensor_enable(temp_sensor));
    float tvalue;
    ESP_ERROR_CHECK(temperature_sensor_get_celsius(temp_sensor, &tvalue));
    ESP_ERROR_CHECK(temperature_sensor_disable(temp_sensor));
    return mp_obj_new_float((mp_float_t)tvalue);
}
STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(esp32_raw_temperature_obj, 0, 2, esp32_raw_temperature);
```
The settings would be effective only on the first call to raw_temperature. In my tests, it did not make any difference which range was chosen. Even setting min larger than max did not make a difference.

**Reply by robert-hh** [reply]
Scanning through the esp source code, setting the range is waste of code, since it will be adapted automatically depending on the actual value.

**Comment by rini711** [comment]
Thank you @robert-hh for the Support , I have added the above code in modesp32.c file and get the correct value of temperature. But after hard resetting the device, first time temperature reading is coming incorrectly. 

First time temperature value is coming as more than 60. Is there anyway to fix this issue?

Please find the below screenshot

<img width="349" alt="image" src="https://github.com/micropython/micropython/assets/144786971/6d462d8a-6b1d-4086-ad2a-acfe3198ab54">


Code Added in modesp32.c file

#if CONFIG_IDF_TARGET_ESP32C3
#include "driver/temperature_sensor.h"

STATIC mp_obj_t esp32_raw_temperature(void) {
    static temperature_sensor_handle_t temp_sensor = NULL;
    if (temp_sensor == NULL) {
        temperature_sensor_config_t temp_sensor_config = TEMPERATURE_SENSOR_CONFIG_DEFAULT(20, 100);
        ESP_ERROR_CHECK(temperature_sensor_install(&temp_sensor_config, &temp_sensor));
    }
    ESP_ERROR_CHECK(temperature_sensor_enable(temp_sensor));
    float tvalue;
    ESP_ERROR_CHECK(temperature_sensor_get_celsius(temp_sensor, &tvalue));
    ESP_ERROR_CHECK(temperature_sensor_disable(temp_sensor));
    return mp_obj_new_int((int)(tvalue+0.5));
}
STATIC MP_DEFINE_CONST_FUN_OBJ_0(esp32_raw_temperature_obj, esp32_raw_temperature);

#endif

**Reply by robert-hh** [reply]
I cannot replicate that behavior here. The first value is always reasonable. Since you use Thonny, please try the same without Thonny using a simple reliable terminal emulator like Putty and paste the above code to REPL with the sequence:

Ctrl-E
<paste the code>
Ctrl-D

You could as well upload that test code as a script to the board and run from REPL using e.g. Putty:

import <test_scrip>

**Reply by robert-hh** [reply]
Update: I can replicate that behavior by adding a call to `temperature_sensor_uninstall(temp_sensor)` to be executed on soft reset. Then I get a sequence of falling temp values until it reaches a certain level. That does not happen after a power cycle, hard reset or machine.reset() call. This sequence does not depend on the waiting time after soft reset. And it does definitely not reflect the real temperature of the device. So maybe a bug in the idf.
```
60
41
39
37
36
35
35
34
33
33
```

**Reply by robert-hh** [reply]
Clearing a static symbol in temperature_sensor_uninstall() in the esp_idf I can reduce the number of wrong values to 1. That could be then dropped by reading the temperature twice on the first call and dropping the first value. No good style though.
So better do not uninstall the sensor on soft reset.

**Comment by robert-hh** [comment]
b.t.w: Which esp-idf version do you use, v5.0.2 or v5.0.4? There are quite a few changes to the temperature sensor code between the two versions. The actual main line version is v5.0.4.

**Comment by ricksorensen** [comment]
I am using 5.0.4 currently ... but 5.0.2 seemed to work also ... maybe I
better check!
I am seeing what seems like a single spurious read after install --- but
today the values I read are 4deg C less than yesterday.  Curious because it
is actually a bit warmer today externally.  Maybe the range settings to
have an impact sometimes?

On Wed, Dec 13, 2023 at 9:51 AM Robert Hammelrath ***@***.***>
wrote:

> b.t.w: Which esp-idf version do you use, v5.0.2 or v5.0.4? There are quite
> a few changes to the temperature sensor code between the two versions. The
> actual main line version is v5.0.4.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/micropython/discussions/10443#discussioncomment-7844122>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA5QPQA6CNQE5WGKE7UFCLDYJHFGPAVCNFSM6AAAAAATUSXAISVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TQNBUGEZDE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
Rick Sorensen
***@***.***
651-470-5511

**Comment by robert-hh** [comment]
The range setting will be changed by the esp-idf code depending on the actual temperature reading. The range at calling temperature_sensor_install() is just the starting point, and something in the middle range like 10,50 or 20,100 is a good starting choice.
My tests show, that temperature_sensor_uninstall() leaves the esp-idf driver in an inconsistent state, which will cause the readings after another temperature_sensor_install() to be wrong, at least the first reading. That's what caused the initial code to fail, which in every loop called install/uninstall. With that, only the first call after reset was right. So better just call temperature_sensor_install() once.

**Reply by robert-hh** [reply]
This is my updated modesp32.c file, including support for ESP32S2 and ESP32S3. 

[modesp32.zip](https://github.com/micropython/micropython/files/13664756/modesp32.zip)

**Comment by ricksorensen** [comment]
Thanks --- looks good.  I had come to pretty much the same start up.  I did
add some test functions for playing with level (low and high) settings and
uninstall too ... but I think except for extreme low power requirements
(which probably should not heat up cpu much) an uninstall is may not be
needed.

Hope you get time this holiday to warm your fingers.

On Wed, Dec 13, 2023 at 1:21 PM Robert Hammelrath ***@***.***>
wrote:

> This is my updated modesp32.c file, including support for ESP32S2 and
> ESP32S3.
>
> modesp32.zip
> <https://github.com/micropython/micropython/files/13664756/modesp32.zip>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/micropython/discussions/10443#discussioncomment-7846335>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA5QPQBBFGGCFEHPL2UAYTTYJH53NAVCNFSM6AAAAAATUSXAISVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TQNBWGMZTK>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
Rick Sorensen
***@***.***
651-470-5511

**Comment by robert-hh** [comment]
Thanks. I compared a few C3 boards for their start temperature after having them settled together w/o power. There are differences of 2 °C (max - min) of the start value, which is about the room temperature. But the power consumption between boards of the same type varies as well slightly. Then ones with a higher start temperature and faster ramp up consume slightly more current. So it looks fine.

**Reply by robert-hh** [reply]
@ricksorensen Do you want to make a PR for it? 
There is still an inconsistency now between the ESP32 and the other MCUs. For ESP32, the return value is really a "raw" quantity with unknown physical unit and scale. For the other MCUs its not raw, but a degree Celsius value. So the name "raw_temperature" of the method is misleading. Maybe it should better be "mcu_temperature" or just "temperature".

**Comment by ricksorensen** [comment]
I will attempt that this weekend.

On Thu, Dec 14, 2023 at 1:50 AM Robert Hammelrath ***@***.***>
wrote:

> @ricksorensen <https://github.com/ricksorensen> Do you want to make a PR
> for it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/micropython/discussions/10443#discussioncomment-7850639>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA5QPQDSM23OTUSXX7LKUODYJKVVJAVCNFSM6AAAAAATUSXAISVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TQNJQGYZTS>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
Rick Sorensen
***@***.***
651-470-5511

**Reply by robert-hh** [reply]
OK. This is the actual state of the file. I renamed "raw_temperature" to "temperature" for the new supported boards. The documentation has to be adapted.
Edit: you might rename it back to raw_temperature, but there must be a few words in the documentation about the difference between the MCU families.

[modesp32.zip](https://github.com/micropython/micropython/files/13673637/modesp32.zip)

**Reply by robert-hh** [reply]
I made a fast test with an ESP32C and two ESP32 starting up at a) room 21°C, b) kitchen fridge at -18°C, c) kitchen stove at 65° C. Temperatures checked with a DS18B20. The ESP32 devices were taken out of the fridge/oven for testing, with like  30 seconds between taking out and test start. The ESP32C3 device stayed in the fridge during test. Results:
```
        ESP32C3  ESP32S2_1  ESP32S2_2  ESP32S3  ESP32_1 ESP32_2
-18°C    -17      -32        -18        -20      100     -
 21°C     22       17         19         20      114     100
 60°C     62       54         59         61      122     112
```
The numbers are the first values after boot. The values in brackets for the ESP32S2_1 are those after about 20 seconds operation.

Keeping in mind the low thermal capacity of the devices, the heat up during device start and the +/- 1 error of digital instruments, the results for the ESP32C3, ESP32S2_2 and ESP32S3 are very good. The ESP32S2_1 is far too low of at low and high temperatures, which does not change when the default range is adapted. The ESP32 results are hardly useful. That may be the reason why espressif dropped support for it.

**Reply by ricksorensen** [reply]
How about cpu_temperature?

**Reply by robert-hh** [reply]
Whatever you prefer. I had considered mcu_temperature as well, but went for the shorter word.

**Reply by robert-hh** [reply]
In my last (not sent) version I renamed the function & object for mcu_temperature to esp32_mcu_temperature() and esp32_mcu_temperature_obj as well. This is just internal stuff in modesp32.c, so no need to change it. Sorry for not mentioning before.
```
STATIC mp_obj_t esp32_mcu_temperature(void) {
    static temperature_sensor_handle_t temp_sensor = NULL;
    float tvalue;
    if (temp_sensor == NULL) {
        temperature_sensor_config_t temp_sensor_config = TEMPERATURE_SENSOR_CONFIG_DEFAULT(0, 80);
        ESP_ERROR_CHECK(temperature_sensor_install(&temp_sensor_config, &temp_sensor));
    }
    ESP_ERROR_CHECK(temperature_sensor_enable(temp_sensor));
    ESP_ERROR_CHECK(temperature_sensor_get_celsius(temp_sensor, &tvalue));
    ESP_ERROR_CHECK(temperature_sensor_disable(temp_sensor));
    return mp_obj_new_int((int)(tvalue + 0.5));
}
STATIC MP_DEFINE_CONST_FUN_OBJ_0(esp32_mcu_temperature_obj, esp32_mcu_temperature);
```
and
```
    { MP_ROM_QSTR(MP_QSTR_mcu_temperature), MP_ROM_PTR(&esp32_mcu_temperature_obj) },
```

**Comment by ricksorensen** [comment]
Once I figure out how to correct the format on the commit message I'll
update the request

**Reply by robert-hh** [reply]
It complained about the lines being too long. you can change previous commits with `rebase -i <commit-hash>`. See the git docs for git rebase.
Edit: Tools like `sublime-merge` allow to directly edit commit messages. Editing the commit content with that tool is possible as well, but it is more clear to use the command line version instead.

When you do a `git log --oneline` in your branch, you get something like:
```
0ff60bc5b (HEAD -> esp32xn_temp) esp32/modesp32: Support reading chip temperature on ESP32C3/S2/S3.
755dee14c esp32/quickref.rst: Correct spelling.
3f1975fdb esp32/quickref.rst: Add documentation for new mcu_temperature().
05d3b2230 (origin/master, origin/HEAD) docs/library: Document SSLContext cert methods and asyncio support.
bfd6ad94f extmod/asyncio: Add ssl support with SSLContext.
```
So if you want to change the commit message of `3f1975fdb`, run `git rebase -i 05d3b2230`. Then you get in the editor an action  list of commits:
```
pick 3f1975fdb esp32/quickref.rst: Add documentation for new mcu_temperature().
pick 755dee14c esp32/quickref.rst: Correct spelling.
pick 0ff60bc5b esp32/modesp32: Support reading chip temperature on ESP32C3/S2/S3.
```
Note the the ordering of lines is reversed. Since you want to change commit `3f1975fdb`, replace the word `pick` by `edit`, and save & store that action list. Then, your branch is rolled back to commit `3f1975fdb`. To just change the commit message, use `git commit --amend`. Git will open an editor with the commit message, which you can edit & save. Then run `git rebase --continue`.
If you just want to change the commit message, you can use the action `reword` instead of `edit`, which brings you immediately into the editor to change the message. That includes automatically the steps of `git commit --amend` and `git rebase --continue`.

If it is just the last commit, which you want to change, you can edit the source files of the commit, save them and run 'git commit -a --amend'. The `-a` will pick up the changed files and the `--amend` allows to edit the commit message.

Note: With this action list of commits you can as well re-order commits (just move the lines) or combine two commits into one (action `squash`).

**Reply by robert-hh** [reply]
@ricksorensen The above table with test results is updated. I added results for ESP32S2 and ESP32S3 boards.
